File: Ferramentas/REPORT SISMOGRAFIA/docs/js/parser.js
/* Parser para CSVs ISEE (Micromate / Instantel).
 * Espelha src/parser.py do pipeline Python: lê o cabeçalho chave→valor
 * até encontrar a tabela de onda (linhas com "Tran, Vert, Long, MicL")
 * e extrai os campos técnicos. Roda 100% no navegador.
 */
(function () {
  'use strict';

  function stripAccents(s) {
    return s.normalize('NFD').replace(/[̀-ͯ]/g, '');
  }
  function normKey(s) {
    return stripAccents(String(s || '')).toLowerCase().replace(/[^a-z0-9]+/g, '');
  }
  function parseFloatBR(v) {
    if (v == null) return null;
    var txt = String(v).replace(/ /g, ' ').trim();
    if (!txt) return null;
    var m = txt.match(/[-+]?\d+(?:[.,]\d+)?/);
    if (!m) return null;
    var n = parseFloat(m[0].replace(',', '.'));
    return isNaN(n) ? null : n;
  }
  function parseScaledDistance(value) {
    var res = { scaled_distance: null, distance_m: null, charge_kg: null };
    if (!value) return res;
    var nums = String(value).match(/[-+]?\d+(?:[.,]\d+)?/g);
    if (!nums) return res;
    nums = nums.map(function (n) { return parseFloat(n.replace(',', '.')); });
    if (nums.length >= 1) res.scaled_distance = nums[0];
    if (nums.length >= 2) res.distance_m = nums[1];
    if (nums.length >= 3) res.charge_kg = nums[2];
    return res;
  }

  function parseRawCsv(text) {
    // Papa parses quoted CSV correctly; header is a series of 2-column rows.
    var out = Papa.parse(text, { skipEmptyLines: true });
    return out.data || [];
  }

  function readHeader(rows) {
    var meta = {};
    var titleNotes = {}, titleStrings = {};
    for (var i = 0; i < rows.length; i++) {
      var row = rows[i].map(function (c) { return (c == null ? '' : String(c)).trim(); });
      if (row.length === 0) continue;
      if (row.length >= 4 &&
          normKey(row[0]) === 'tran' &&
          normKey(row[1]) === 'vert' &&
          normKey(row[2]) === 'long' &&
          normKey(row[3]) === 'micl') {
        break;
      }
      if (row.length >= 2) {
        var key = row[0], val = row[1];
        if (key) {
          meta[key] = val;
          var nk = normKey(key);
          var mNote = nk.match(/^titlenote(\d+)$/);
          var mStr  = nk.match(/^titlestring(\d+)$/);
          if (mNote) titleNotes[mNote[1]] = val;
          if (mStr) titleStrings[mStr[1]] = val;
        }
      }
    }
    Object.keys(titleNotes).forEach(function (idx) {
      var note = titleNotes[idx];
      var val = titleStrings[idx] || '';
      if (note) meta['Title::' + note] = val;
    });
    return meta;
  }

  function parseSismoCsv(fileName, text) {
    var rows = parseRawCsv(text);
    var meta = readHeader(rows);
    var norm = {};
    Object.keys(meta).forEach(function (k) { norm[normKey(k)] = meta[k]; });

    var location = norm['titlelocation'] || meta['Title::Location'] || meta['TitleString1'] || fileName.replace(/\.[^.]+$/, '');
    var client   = norm['titleclient']   || meta['Title::Client']   || meta['TitleString2'] || null;
    var company  = norm['titlecompany']  || meta['Title::Company']  || meta['TitleString3'] || null;

    var scaled = parseScaledDistance(meta['ScaledDistance'] || '');
    var gpsDist = parseFloatBR(meta['GpsDistance']);
    if (gpsDist == null) gpsDist = scaled.distance_m;

    var eventDate = meta['EventDate'] || null;
    var eventTime = meta['EventTime'] || null;

    return {
      source_file: fileName,
      event_date: eventDate,
      event_time: eventTime,
      event_iso: buildIso(eventDate, eventTime),
      point_name: String(location).trim() || fileName.replace(/\.[^.]+$/, ''),
      client: client,
      company: company,
      serial_number: meta['SerialNumber'] || null,
      calibration: meta['Calibration'] || null,
      gps_distance_m: gpsDist,
      scaled_distance: scaled.scaled_distance,
      charge_kg: scaled.charge_kg,
      pspl_db: parseFloatBR(meta['MicPSPL']),
      mic_freq_hz: parseFloatBR(meta['MicZCFreq']),
      pvs_mm_s: parseFloatBR(meta['PeakVectorSum']),
      tran_ppv_mm_s: parseFloatBR(meta['TranPPV']),
      vert_ppv_mm_s: parseFloatBR(meta['VertPPV']),
      long_ppv_mm_s: parseFloatBR(meta['LongPPV']),
      tran_freq_hz: parseFloatBR(meta['TranZCFreq']),
      vert_freq_hz: parseFloatBR(meta['VertZCFreq']),
      long_freq_hz: parseFloatBR(meta['LongZCFreq']),
      mic_test_result: meta['MicTestResults'] || null,
      tran_test_result: meta['TranTestResults'] || null,
      vert_test_result: meta['VertTestResults'] || null,
      long_test_result: meta['LongTestResults'] || null
    };
  }

  function buildIso(date, time) {
    if (!date) return null;
    // Aceita formatos "2026-06-30" ou "30/06/2026" ou "June 30, 2026"
    var iso;
    var m1 = /^(\d{4})-(\d{2})-(\d{2})$/.exec(date);
    var m2 = /^(\d{2})\/(\d{2})\/(\d{4})$/.exec(date);
    if (m1) iso = m1[1] + '-' + m1[2] + '-' + m1[3];
    else if (m2) iso = m2[3] + '-' + m2[2] + '-' + m2[1];
    else {
      var d = new Date(date);
      if (!isNaN(d.getTime())) {
        iso = d.getFullYear() + '-' + String(d.getMonth()+1).padStart(2,'0') + '-' + String(d.getDate()).padStart(2,'0');
      } else {
        return null;
      }
    }
    if (time && /^\d{2}:\d{2}(:\d{2})?$/.test(time)) {
      iso += 'T' + (time.length === 5 ? time + ':00' : time);
    } else {
      iso += 'T00:00:00';
    }
    return iso;
  }

  window.SismoParser = {
    parseSismoCsv: parseSismoCsv,
    normKey: normKey,
    parseFloatBR: parseFloatBR
  };
})();

File: Ferramentas/REPORT SISMOGRAFIA/src/parser.py
from __future__ import annotations

import csv
import re
import unicodedata
from dataclasses import dataclass, asdict
from pathlib import Path
from typing import Any, Dict, List, Optional


def _strip_accents(value: str) -> str:
    return ''.join(c for c in unicodedata.normalize('NFD', value) if unicodedata.category(c) != 'Mn')


def norm_key(value: str) -> str:
    return re.sub(r'[^a-z0-9]+', '', _strip_accents(str(value)).lower())


def parse_float(value: Any) -> Optional[float]:
    if value is None:
        return None
    txt = str(value).strip().replace('\u00a0', ' ')
    if not txt:
        return None
    # Captura o primeiro número em formatos 1.234,56 ou 1234.56.
    m = re.search(r'[-+]?\d+(?:[\.,]\d+)?', txt)
    if not m:
        return None
    number = m.group(0).replace(',', '.')
    try:
        return float(number)
    except ValueError:
        return None


def parse_scaled_distance(value: str) -> Dict[str, Optional[float]]:
    """Extrai SD, distância e carga de textos como '4586.9 (1450.5 m, 0.1 kg)'."""
    result = {"scaled_distance": None, "distance_m": None, "charge_kg": None}
    if not value:
        return result
    nums = re.findall(r'[-+]?\d+(?:[\.,]\d+)?', str(value))
    nums = [float(n.replace(',', '.')) for n in nums]
    if len(nums) >= 1:
        result["scaled_distance"] = nums[0]
    if len(nums) >= 2:
        result["distance_m"] = nums[1]
    if len(nums) >= 3:
        result["charge_kg"] = nums[2]
    return result


def find_input_csvs(input_path: str | Path) -> List[Path]:
    path = Path(input_path)
    if path.is_file():
        return [path]
    if not path.exists():
        raise FileNotFoundError(f"Entrada não encontrada: {path}")
    files = sorted([p for p in path.rglob('*') if p.suffix.lower() == '.csv'])
    if not files:
        raise FileNotFoundError(f"Nenhum CSV encontrado em: {path}")
    return files


@dataclass
class SismoRecord:
    source_file: str
    event_date: Optional[str]
    event_time: Optional[str]
    point_name: str
    client: Optional[str]
    company: Optional[str]
    serial_number: Optional[str]
    calibration: Optional[str]
    gps_distance_m: Optional[float]
    scaled_distance: Optional[float]
    charge_kg: Optional[float]
    pspl_db: Optional[float]
    mic_freq_hz: Optional[float]
    pvs_mm_s: Optional[float]
    tran_ppv_mm_s: Optional[float]
    vert_ppv_mm_s: Optional[float]
    long_ppv_mm_s: Optional[float]
    tran_freq_hz: Optional[float]
    vert_freq_hz: Optional[float]
    long_freq_hz: Optional[float]
    tran_time_peak_s: Optional[float]
    vert_time_peak_s: Optional[float]
    long_time_peak_s: Optional[float]
    mic_time_peak_s: Optional[float]
    mic_test_result: Optional[str]
    tran_test_result: Optional[str]
    vert_test_result: Optional[str]
    long_test_result: Optional[str]
    metadata: Dict[str, str]

    def to_dict(self) -> Dict[str, Any]:
        return asdict(self)


def read_header_metadata(path: str | Path) -> Dict[str, str]:
    """Lê apenas o cabeçalho do CSV até encontrar a tabela de onda (Tran, Vert, Long, MicL)."""
    metadata: Dict[str, str] = {}
    path = Path(path)
    with path.open('r', encoding='utf-8-sig', newline='', errors='replace') as f:
        reader = csv.reader(f)
        title_notes: Dict[str, str] = {}
        title_strings: Dict[str, str] = {}
        for row in reader:
            if not row:
                continue
            compact = [c.strip() for c in row]
            if len(compact) >= 4 and [norm_key(c) for c in compact[:4]] == ['tran', 'vert', 'long', 'micl']:
                break
            if len(compact) >= 2:
                key, value = compact[0], compact[1]
                metadata[key] = value
                nk = norm_key(key)
                note_match = re.fullmatch(r'titlenote(\d+)', nk)
                string_match = re.fullmatch(r'titlestring(\d+)', nk)
                if note_match:
                    title_notes[note_match.group(1)] = value
                if string_match:
                    title_strings[string_match.group(1)] = value
        for idx, note in title_notes.items():
            value = title_strings.get(idx, '')
            if note:
                metadata[f'Title::{note}'] = value
    return metadata


def parse_sismo_csv(path: str | Path) -> SismoRecord:
    path = Path(path)
    meta = read_header_metadata(path)
    normalized = {norm_key(k): v for k, v in meta.items()}

    location = normalized.get('titlelocation') or meta.get('Title::Location') or meta.get('TitleString1') or path.stem
    client = normalized.get('titleclient') or meta.get('Title::Client') or meta.get('TitleString2')
    company = normalized.get('titlecompany') or meta.get('Title::Company') or meta.get('TitleString3')

    scaled = parse_scaled_distance(meta.get('ScaledDistance', ''))
    gps_distance = parse_float(meta.get('GpsDistance')) or scaled.get('distance_m')

    return SismoRecord(
        source_file=path.name,
        event_date=meta.get('EventDate'),
        event_time=meta.get('EventTime'),
        point_name=str(location).strip() or path.stem,
        client=client,
        company=company,
        serial_number=meta.get('SerialNumber'),
        calibration=meta.get('Calibration'),
        gps_distance_m=gps_distance,
        scaled_distance=scaled.get('scaled_distance'),
        charge_kg=scaled.get('charge_kg'),
        pspl_db=parse_float(meta.get('MicPSPL')),
        mic_freq_hz=parse_float(meta.get('MicZCFreq')),
        pvs_mm_s=parse_float(meta.get('PeakVectorSum')),
        tran_ppv_mm_s=parse_float(meta.get('TranPPV')),
        vert_ppv_mm_s=parse_float(meta.get('VertPPV')),
        long_ppv_mm_s=parse_float(meta.get('LongPPV')),
        tran_freq_hz=parse_float(meta.get('TranZCFreq')),
        vert_freq_hz=parse_float(meta.get('VertZCFreq')),
        long_freq_hz=parse_float(meta.get('LongZCFreq')),
        tran_time_peak_s=parse_float(meta.get('TranTimeofPeak')),
        vert_time_peak_s=parse_float(meta.get('VertTimeofPeak')),
        long_time_peak_s=parse_float(meta.get('LongTimeofPeak')),
        mic_time_peak_s=parse_float(meta.get('MicTimeofPeak')),
        mic_test_result=meta.get('MicTestResults'),
        tran_test_result=meta.get('TranTestResults'),
        vert_test_result=meta.get('VertTestResults'),
        long_test_result=meta.get('LongTestResults'),
        metadata=meta,
    )


def parse_many(input_path: str | Path) -> List[SismoRecord]:
    return [parse_sismo_csv(p) for p in find_input_csvs(input_path)]
